File: testers/avro/java/generated-and-checked-in/com/example/events/precisetypes/Decimal10_2.java
package com.example.events.precisetypes;

import dev.typr.foundations.data.precise.DecimalN;
import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.Optional;

public record Decimal10_2(BigDecimal value) implements DecimalN {
  @java.lang.Deprecated
  public Decimal10_2 {}

  public Decimal10_2 withValue(BigDecimal value) {
    return new Decimal10_2(value);
  }

  @Override
  public java.lang.String toString() {
    return value.toString();
  }

  public static Optional<Decimal10_2> of(BigDecimal value) {
    BigDecimal scaled = value.setScale(2, RoundingMode.HALF_UP);
    return scaled.precision() <= 10 ? Optional.of(new Decimal10_2(scaled)) : Optional.empty();
  }

  public static Decimal10_2 of(Integer value) {
    return new Decimal10_2(BigDecimal.valueOf((long) (value)));
  }

  public static Optional<Decimal10_2> of(Long value) {
    return Decimal10_2.of(BigDecimal.valueOf(value));
  }

  public static Optional<Decimal10_2> of(Double value) {
    return Decimal10_2.of(BigDecimal.valueOf(value));
  }

  public static Decimal10_2 unsafeForce(BigDecimal value) {
    BigDecimal scaled = value.setScale(2, RoundingMode.HALF_UP);
    if (scaled.precision() > 10) {
      throw new IllegalArgumentException("Value exceeds precision(10, 2)");
    }
    ;
    return new Decimal10_2(scaled);
  }

  @Override
  public BigDecimal decimalValue() {
    return value;
  }

  @Override
  public boolean equals(Object obj) {
    if (this == obj) return true;
    if (!(obj instanceof DecimalN other)) return false;
    return decimalValue().compareTo(other.decimalValue()) == 0;
  }

  @Override
  public int hashCode() {
    return decimalValue().stripTrailingZeros().hashCode();
  }

  @Override
  public int precision() {
    return 10;
  }

  @Override
  public int scale() {
    return 2;
  }

  @Override
  public boolean semanticEquals(DecimalN other) {
    return (other == null ? false : decimalValue().compareTo(other.decimalValue()) == 0);
  }

  @Override
  public int semanticHashCode() {
    return decimalValue().stripTrailingZeros().hashCode();
  }
}

File: testers/avro/kotlin-json/generated-and-checked-in/com/example/events/precisetypes/Decimal10_2.kt
package com.example.events.precisetypes

import dev.typr.foundations.data.precise.DecimalN
import java.lang.IllegalArgumentException
import java.math.BigDecimal
import java.math.RoundingMode

@kotlin.ConsistentCopyVisibility
data class Decimal10_2 private constructor(val value: BigDecimal) : DecimalN {
  override fun decimalValue(): BigDecimal = value

  override fun equals(other: Any?): kotlin.Boolean {
    if (this === other) return true
    if (other !is DecimalN) return false
    return decimalValue().compareTo(other.decimalValue()) == 0
  }

  override fun hashCode(): Int = decimalValue().stripTrailingZeros().hashCode()

  override fun precision(): Int = 10

  override fun scale(): Int = 2

  override fun semanticEquals(other: DecimalN): kotlin.Boolean = if (other == null) false else decimalValue().compareTo(other.decimalValue()) == 0

  override fun semanticHashCode(): Int = decimalValue().stripTrailingZeros().hashCode()

  override fun toString(): kotlin.String {
    return value.toString()
  }

  companion object {
    fun of(value: BigDecimal): Decimal10_2? {
      val scaled = value.setScale(2, RoundingMode.HALF_UP)
      return if (scaled.precision() <= 10) Decimal10_2(scaled) else null
    }

    fun of(value: Int): Decimal10_2 = Decimal10_2(BigDecimal.valueOf(value.toLong()))

    fun of(value: kotlin.Long): Decimal10_2? = Decimal10_2.of(BigDecimal.valueOf(value))

    fun of(value: kotlin.Double): Decimal10_2? = Decimal10_2.of(BigDecimal.valueOf(value))

    fun unsafeForce(value: BigDecimal): Decimal10_2 {
      val scaled = value.setScale(2, RoundingMode.HALF_UP)
      if (scaled.precision() > 10) throw IllegalArgumentException("Value exceeds precision(10, 2)")
      return Decimal10_2(scaled)
    }
  }
}